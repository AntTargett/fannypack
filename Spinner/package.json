{
  "name": "fannypack/Spinner",
  "private": true,
  "main": "../lib/Spinner/index.js",
  "module": "../es/Spinner/index.js"
}