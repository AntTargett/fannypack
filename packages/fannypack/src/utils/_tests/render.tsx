import * as React from 'react';
import { render } from '@testing-library/react';
// @ts-ignore
import { Provider } from '../../Provider';

export default (Component: any, { theme = {} } = {}) => render(<Provider theme={theme}>{Component}</Provider>);
