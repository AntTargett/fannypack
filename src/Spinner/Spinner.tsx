import * as React from 'react';
import * as PropTypes from 'prop-types';
import { BoxProps as ReakitBoxProps } from '@jmoxey/reakit/ts/Box/Box';

import { Omit } from '../types';
import _Spinner from './styled';

export type LocalSpinnerProps = {
  className?: string;
  color?: string;
  size?: string;
};
export type SpinnerProps = Omit<ReakitBoxProps, 'children'> & LocalSpinnerProps;

export const Spinner: React.FunctionComponent<LocalSpinnerProps> = ({ className, color, size, ...props }) => (
  <_Spinner className={className} color={color} styledSize={size} {...props} />
);

Spinner.propTypes = {
  className: PropTypes.string,
  color: PropTypes.string,
  size: PropTypes.string
};
Spinner.defaultProps = {
  className: undefined,
  color: 'primary',
  size: 'default'
};

export default Spinner;
