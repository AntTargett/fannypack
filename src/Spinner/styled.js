import styled, { css, keyframes } from 'reakit/styled';
import { palette, theme } from 'styled-tools';
import { Box } from '../primitives';

const sizes = {
  small: css`
    & {
      font-size: 14px;
      border-width: 2px;
    }
    & {
      ${theme('fannypack.Spinner.sizes.small')};
    }
  `,
  medium: css`
    & {
      font-size: 28px;
    }
    & {
      ${theme('fannypack.Spinner.sizes.medium')};
    }
  `,
  large: css`
    & {
      font-size: 36px;
    }
    & {
      ${theme('fannypack.Spinner.sizes.large')};
    }
  `
};

const rotate = keyframes`
  from {
    transform: rotate(0deg);
  }

  to {
    transform: rotate(360deg);
  }
`;
export const spinnerProperties = css`
  animation: ${rotate} 0.6s infinite linear;
  border: 0.1em solid ${props => palette(props.color)(props)};
  border-radius: 100%;
  border-right-color: transparent;
  border-top-color: transparent;
  display: inline-block;
  height: 1em;
  width: 1em;
  position: relative;
`;

export default styled(Box)`
  font-size: 20px;
  ${spinnerProperties};

  & {
    ${theme('fannypack.Spinner.base')};
  }

  ${props => sizes[props.size]} /* Extend size styles */;
`;
