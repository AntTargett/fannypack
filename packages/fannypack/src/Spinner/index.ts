import Spinner from './Spinner';

export default Spinner;

export * from './Spinner';
export { default as Spinner } from './Spinner';
