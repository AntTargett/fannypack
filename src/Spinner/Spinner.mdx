---
name: Spinner
route: /components/spinner
menu: Components
---

import { Playground } from 'docz';
import PropsTable from '../_docs/components/PropsTable';
import Button from '../Button';
import {Spinner} from './Spinner';

# Spinners

## Import

`import { Spinner } from 'fannypack'`

## Basic Usage

<Playground>
  <Spinner />
</Playground>

## Sizes

<Playground>
  <Spinner size="small" />
  <Spinner marginLeft="2u" />
  <Spinner size="medium" marginLeft="2u" />
  <Spinner size="large" marginLeft="2u" />
</Playground>

## Colors

<Playground>
  <Spinner color="success" />
  <Spinner color="danger" marginLeft="2u" />
  <Spinner color="warning" marginLeft="2u" />
</Playground>


## Props

<PropsTable component={Spinner} />

## Theming

### Schema

```jsx
{
  Spinner: {
    base: css | Object,
    sizes: {
      small: css | Object,
      medium: css | Object,
      large: css | Object
    }
  }
}
```
