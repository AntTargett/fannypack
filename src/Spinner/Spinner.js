import React from 'react';
import PropTypes from 'prop-types';

import { Spinner as StyledSpinner } from './styled';

const Spinner = ({ className, color, size, state }) => (
  <StyledSpinner
    className={className}
    color={color}
    palette={state}
    size={size}
    version="1.1"
    xmlns="http://www.w3.org/2000/svg"
    x="0px"
    y="0px"
    viewBox="0 0 50 50"
    xmlSpace="preserve"
  >
    <path d="M43.935,25.145c0-10.318-8.364-18.683-18.683-18.683c-10.318,0-18.683,8.365-18.683,18.683h4.068c0-8.071,6.543-14.615,14.615-14.615c8.072,0,14.615,6.543,14.615,14.615H43.935z">
      <animateTransform
        attributeType="xml"
        attributeName="transform"
        type="rotate"
        from="0 25 25"
        to="360 25 25"
        dur="0.6s"
        repeatCount="indefinite"
      />
    </path>
  </StyledSpinner>
);

Spinner.propTypes = {
  className: PropTypes.string,
  /** Color of the loading spinner. */
  color: PropTypes.string,
  /** Size of the loading spinner. */
  size: PropTypes.oneOf(['small', 'default', 'medium', 'large']),
  /** State of the loading spinner. */
  state: PropTypes.oneOf(['default', 'primary', 'secondary', 'success', 'danger', 'warning'])
};

Spinner.defaultProps = {
  className: null,
  color: null,
  state: 'primary',
  size: 'default'
};

export default Spinner;
