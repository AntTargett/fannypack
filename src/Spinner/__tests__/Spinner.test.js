import React from 'react';
import render from '../../utils/tests/render';
import Spinner from '../Spinner';
import 'jest-styled-components';

it('renders correctly', () => {
  const { container } = render(<Spinner />);
  expect(container.firstChild).toMatchSnapshot();
});

it('renders correctly when a color prop is set', () => {
  const { container } = render(<Spinner color="red" />);
  expect(container.firstChild).toMatchSnapshot();
});

it('renders correctly when a size prop is set', () => {
  const { container } = render(<Spinner size="large" />);
  expect(container.firstChild).toMatchSnapshot();
});

it('renders correctly when a palette prop is set', () => {
  const { container } = render(<Spinner palette="primary" />);
  expect(container.firstChild).toMatchSnapshot();
});
