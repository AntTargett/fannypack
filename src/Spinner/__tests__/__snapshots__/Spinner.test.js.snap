// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
.c1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

.c0 {
  font-family: system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Fira Sans', 'Droid Sans','Helvetica Neue',sans-serif;
  font-size: 16px;
  font-weight: 400;
  margin: 0;
  padding: 0;
  -webkit-font-smoothing: antialiased;
  text-rendering: optimizeLegibility;
}

.c0 *:focus {
  outline: 2px solid #112ebb;
  outline-offset: 2px;
}

.c2 {
  width: 1.5rem;
  height: 1.5rem;
}

.c2 path {
  fill: #112ebb;
}

<div
  class="c0 c1"
>
  <svg
    class="c2"
    color=""
    version="1.1"
    viewBox="0 0 50 50"
    x="0px"
    xml:space="preserve"
    xmlns="http://www.w3.org/2000/svg"
    y="0px"
  >
    <path
      d="M43.935,25.145c0-10.318-8.364-18.683-18.683-18.683c-10.318,0-18.683,8.365-18.683,18.683h4.068c0-8.071,6.543-14.615,14.615-14.615c8.072,0,14.615,6.543,14.615,14.615H43.935z"
    >
      <animatetransform
        attributeName="transform"
        attributeType="xml"
        dur="0.6s"
        from="0 25 25"
        repeatCount="indefinite"
        to="360 25 25"
        type="rotate"
      />
    </path>
  </svg>
</div>
`;

exports[`renders correctly when a color prop is set 1`] = `
.c1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

.c0 {
  font-family: system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Fira Sans', 'Droid Sans','Helvetica Neue',sans-serif;
  font-size: 16px;
  font-weight: 400;
  margin: 0;
  padding: 0;
  -webkit-font-smoothing: antialiased;
  text-rendering: optimizeLegibility;
}

.c0 *:focus {
  outline: 2px solid #112ebb;
  outline-offset: 2px;
}

.c2 {
  width: 1.5rem;
  height: 1.5rem;
}

.c2 path {
  fill: red;
}

<div
  class="c0 c1"
>
  <svg
    class="c2"
    color="red"
    version="1.1"
    viewBox="0 0 50 50"
    x="0px"
    xml:space="preserve"
    xmlns="http://www.w3.org/2000/svg"
    y="0px"
  >
    <path
      d="M43.935,25.145c0-10.318-8.364-18.683-18.683-18.683c-10.318,0-18.683,8.365-18.683,18.683h4.068c0-8.071,6.543-14.615,14.615-14.615c8.072,0,14.615,6.543,14.615,14.615H43.935z"
    >
      <animatetransform
        attributeName="transform"
        attributeType="xml"
        dur="0.6s"
        from="0 25 25"
        repeatCount="indefinite"
        to="360 25 25"
        type="rotate"
      />
    </path>
  </svg>
</div>
`;

exports[`renders correctly when a palette prop is set 1`] = `
.c1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

.c0 {
  font-family: system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Fira Sans', 'Droid Sans','Helvetica Neue',sans-serif;
  font-size: 16px;
  font-weight: 400;
  margin: 0;
  padding: 0;
  -webkit-font-smoothing: antialiased;
  text-rendering: optimizeLegibility;
}

.c0 *:focus {
  outline: 2px solid #112ebb;
  outline-offset: 2px;
}

.c2 {
  width: 1.5rem;
  height: 1.5rem;
}

.c2 path {
  fill: #112ebb;
}

<div
  class="c0 c1"
>
  <svg
    class="c2"
    color=""
    version="1.1"
    viewBox="0 0 50 50"
    x="0px"
    xml:space="preserve"
    xmlns="http://www.w3.org/2000/svg"
    y="0px"
  >
    <path
      d="M43.935,25.145c0-10.318-8.364-18.683-18.683-18.683c-10.318,0-18.683,8.365-18.683,18.683h4.068c0-8.071,6.543-14.615,14.615-14.615c8.072,0,14.615,6.543,14.615,14.615H43.935z"
    >
      <animatetransform
        attributeName="transform"
        attributeType="xml"
        dur="0.6s"
        from="0 25 25"
        repeatCount="indefinite"
        to="360 25 25"
        type="rotate"
      />
    </path>
  </svg>
</div>
`;

exports[`renders correctly when a size prop is set 1`] = `
.c1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

.c0 {
  font-family: system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Fira Sans', 'Droid Sans','Helvetica Neue',sans-serif;
  font-size: 16px;
  font-weight: 400;
  margin: 0;
  padding: 0;
  -webkit-font-smoothing: antialiased;
  text-rendering: optimizeLegibility;
}

.c0 *:focus {
  outline: 2px solid #112ebb;
  outline-offset: 2px;
}

.c2 {
  width: 1.5rem;
  height: 1.5rem;
}

.c2 {
  width: 3rem;
  height: 3rem;
}

.c2 path {
  fill: #112ebb;
}

<div
  class="c0 c1"
>
  <svg
    class="c2"
    color=""
    version="1.1"
    viewBox="0 0 50 50"
    x="0px"
    xml:space="preserve"
    xmlns="http://www.w3.org/2000/svg"
    y="0px"
  >
    <path
      d="M43.935,25.145c0-10.318-8.364-18.683-18.683-18.683c-10.318,0-18.683,8.365-18.683,18.683h4.068c0-8.071,6.543-14.615,14.615-14.615c8.072,0,14.615,6.543,14.615,14.615H43.935z"
    >
      <animatetransform
        attributeName="transform"
        attributeType="xml"
        dur="0.6s"
        from="0 25 25"
        repeatCount="indefinite"
        to="360 25 25"
        type="rotate"
      />
    </path>
  </svg>
</div>
`;
